### PR TITLE
FCREPO-3753

### PR DESCRIPTION
**Pull request by Geoffsc**
This PR resolves Jira ticket FCREPO-3753: adds a generic HTTP message service to the Camel Toolbox

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3753

### CODE REVIEW

**Comment by demiankatz** [comment]
Note that this is very much a minimum viable product -- it simply forwards the JMS messages to HTTP as headers. We couldn't figure out how to reformat them into a message body and ultimately discovered that they work just fine as headers.

I have just completed a test integration with our application using the code in this branch, and it appears to be fully functional.

Note that there are several details discussed on the August 26 meeting which are NOT currently implemented here:

- Support for basic authentication
- Support for addition of arbitrary headers
- Support for routing to more than one HTTP endpoint
- Support for message filtering

The question is whether we can live without some or all of these; I would think that at very least, basic authentication and/or arbitrary headers would be important to support security needs.

Happy to discuss/expand as needed, but for now we'd like a review of what we've done so far before we invest more time into it, just to be sure we're not too far off base.

**Comment by demiankatz** [comment]
Another important detail is reindexing support: this has NOT been tested yet, and I suspect it probably won't work as expected. I'm not sure of the best way to test, but happy to do so when time permits.

File: fcrepo-indexing-http/pom.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.camel</groupId>
+      <artifactId>camel-http</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.camel</groupId>
+      <artifactId>camel-mustache</artifactId>
```
Comment (demiankatz) [inline fcrepo-indexing-http/pom.xml:39]: I suspect that some of these dependencies aren't actually needed here; we based this pom.xml on the Solr component, and there's probably some junk that can be trimmed out.

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import org.fcrepo.camel.common.config.ConditionOnPropertyTrue;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * A configuration class for the Http Indexer service
+ *
+ * @author Geoff Scholl, Demian Katz
```
Comment (demiankatz) [inline fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java:32]: What's the convention for `@author` tags? Is this okay, or should there be a separate line for each author?

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Value("${input.stream:broker:topic:fedora}")
+    private String inputStream;
+
+    @Value("${http.reindex.stream:broker:queue:http.reindex}")
+    private String reindexStream;
+
+    @Value("${filter.containers:http://localhost:8080/fcrepo/rest/audit}")
+    private String filterContainers;
```
Comment (demiankatz) [inline fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java:53]: Not sure what this is for; is it used?

**Review by demiankatz — COMMENTED** [review]
Just a few comments that might be helpful...

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java (hunk 2 above)
Comment (dbernstein) [reply fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java:32]: Separate line I believe is preferable.

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java (hunk 3 above)
Comment (dbernstein) [reply fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java:53]: This is for ignoring messages from resources that start with the specified paths.

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java (hunk 2 above)
Comment (demiankatz) [reply fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java]: Thanks -- fixed!

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/HttpRouter.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            .process(new EventProcessor())
+            .log(LoggingLevel.TRACE, "Received message from Fedora routing to index.http")
+            .to("direct:send.to.http");
+
+        /*
+         * Handle re-index events
+         */
+        from(config.getReindexStream())
+            .routeId("FcrepoHttpReindex")
+            .to("direct:send.to.http");
```
Comment (dbernstein) [inline fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/HttpRouter.java:73]: What happens if you add in

``` .process(new EventProcessor())``` between 71 and 72?

**Comment by demiankatz** [comment]
@dbernstein, I believe this is now fully functional -- seems to be working as expected for both normal Fedora updates and the reindex case. Can you please give it another review and let us know if you see any rough edges that need polishing? We'll see about adding some test coverage tomorrow.

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Value("${http.baseUrl:http://localhost:8983/http/collection1}")
```
Comment (dbernstein) [inline fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java:56]: I would let the default value be blank.  

```
@Value("${http.baseUrl:}")
```

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    static final String HTTP_INDEXER_ENABLED = "http.indexer.enabled";
+
+    static class HttpIndexerEnabled extends ConditionOnPropertyTrue {
+        HttpIndexerEnabled() {
+            super(FcrepoHttpIndexerConfig.HTTP_INDEXER_ENABLED, false);
+        }
+    }
+
+    @Value("${input.stream:broker:topic:fedora}")
```
Comment (dbernstein) [inline fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java:47]: For consistency I would change this key from `input.stream` to `http.input.stream`

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    private String inputStream;
+
+    @Value("${http.reindex.stream:broker:queue:http.reindex}")
+    private String reindexStream;
+
+    @Value("${filter.containers:http://localhost:8080/fcrepo/rest/audit}")
```
Comment (dbernstein) [inline fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java:53]: For consistency I would change this key from `filter.containers` to `http.filter.containers`

**Review by dbernstein — CHANGES_REQUESTED** [review]
Looking good.  As you mentioned, an integration test would be good.  Another useful thing would be to update the ./README.md with the properties and add a blurb about what the HTTP service does.

File: fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java (hunk 6 above)
Comment (demiankatz) [reply fcrepo-indexing-http/src/main/java/org/fcrepo/camel/indexing/http/FcrepoHttpIndexerConfig.java]: Fixed.

**Comment by demiankatz** [comment]
@dbernstein, I'm about out of time, so let me explain where I've left off: I copied the integration tests from the triplestore component and attempted to replace Fuseki references with MockServer. I was hoping to get the point where I could at least see some failing assertions, but I think I'm still missing a few details to get all the parts to play together nicely.

Specific issues:

1.) Right now, the tests are using `server.verify()` to make assertions about calls to the MockServer HTTP endpoint. This does not appear to be working right. I'm not sure if I'm supposed to do this before or after the calls have occurred, or if I'm using entirely the wrong strategy. All I know is that if I comment out the server.verify() calls, the tests actually seem to get further and do more... I can actually see calls being made to the MockServer in the logs when I comment out the `verify` call, but when I include it, things fail earlier. I just left it in the commit for now to show my work so far.

2.) I don't think the test setup is creating events correctly (or else the code isn't handling the full range of possibilities, or some of the configuration is wrong). When messages get sent to the MockServer, they always seem to have a blank id, and they end up defaulting as Update events even when they're supposed to be Deletes. This suggests to me that for some reason or other, the expected headers are not in the expected places, but I'm not familiar with enough of the details to have a strong feeling of where to go first to fix it.

3.) I removed some `await` calls and other details that seemed to be TripleStore-specific... but maybe we need something similar/equivalent to make the timing work right. Again, I have a vague idea of how this works but don't understand enough details to grasp the whole thing just yet.

I hope I've at least gotten the ball rolling, but some assistance would be greatly appreciated. Happy to devote some time to this on Monday if there is time during/after our sprint meeting, if that's the best path forward.

**Comment by demiankatz** [comment]
Also still to do: update the basic authentication work to reflect forthcoming shared logic.

**Comment by demiankatz** [comment]
Thanks to @mikejritter, I've made significant progress here, and the build should now be passing. However, the tests are not yet correct and need a bit more work; I'm hoping perhaps @dbernstein can help.

The issue is that the wrong event URI is getting passed to the HTTP listener, which I'm almost certain is happening because the test environment is not setting the expected `org.fcrepo.jms.eventtype` header, causing the Camel route to always default to "Update." Right now, both of the tests are expecting `Update` URIs (which is why they are passing), but this is incorrect -- they should be expecting `Create` and `Delete` respectively. My problem is that I don't entirely understand the tests, so I'm not sure how to fix this.

I'm also a bit unclear on why we are creating objects in a real Fedora instance but then manually triggering the Camel route; does that really serve any purpose? I think perhaps that was needed for the Triplestore test that I used as a model since the Triplestore route directly interacts with Fedora... but in these instances, it's probably sufficient to send a fake URI, since testing the HTTP service doesn't really require an active Fedora instance, unless we actually want to test the JMS portion of the integration, which does not appear to be happening here.

Hopefully we can discuss all this later today. Once the tests are sorted out, the only outstanding work that I am aware of is the basic auth refactoring.

**Comment by demiankatz** [comment]
@dbernstein, I believe this is now complete -- I saw that you merged #173, so I merged main into this branch and simplified the basic auth support. I also noticed that the checkstyle plugin was disabled here for some reason (copied and pasted from the Solr indexer, which also has it disabled), so I enabled it and fixed all the issues.

**Pull request merged**